jfix-armeria-facade/src/main/kotlin/ru/fix/armeria/facade/webclient/impl/BaseHttpClientBuilderImpl.kt

```diff
@@ -0,0 +1,258 @@
+package ru.fix.armeria.facade.webclient.impl
+
+import ru.fix.armeria.facade.Either
+import com.linecorp.armeria.client.*
+import com.linecorp.armeria.client.endpoint.EndpointGroup
+import com.linecorp.armeria.client.endpoint.EndpointSelectionStrategy
+import com.linecorp.armeria.common.Flags
+import com.linecorp.armeria.common.SessionProtocol
+import com.linecorp.armeria.common.util.EventLoopGroups
+import org.apache.logging.log4j.kotlin.Logging
+import ru.fix.aggregating.profiler.PrefixedProfiler
+import ru.fix.aggregating.profiler.Profiler
+import ru.fix.armeria.aggregating.profiler.ProfiledConnectionPoolListener
+import ru.fix.armeria.commons.AutoCloseableHttpClient
+import ru.fix.armeria.dynamic.request.endpoint.DynamicAddressEndpoints
+import ru.fix.armeria.dynamic.request.endpoint.SocketAddress
+import ru.fix.armeria.facade.retrofit.RetrofitHttpClientBuilder
+import ru.fix.armeria.facade.retrofit.impl.RetrofitHttpClientBuilderImpl
+import ru.fix.armeria.facade.webclient.BaseHttpClientBuilder
+import ru.fix.armeria.facade.webclient.CloseableWebClient
+import ru.fix.armeria.limiter.RateLimitingHttpClient
+import ru.fix.dynamic.property.api.DynamicProperty
+import ru.fix.stdlib.ratelimiter.RateLimitedDispatcher
+import java.net.URI
+import java.time.Duration
+import java.util.function.Function
+import kotlin.system.measureTimeMillis
+
+internal data class BaseHttpClientBuilderState(
+    val clientNameCreator: () -> String = {
+        throw IllegalStateException("clientName must be set")
+    },
+    val sessionProtocol: SessionProtocol = SessionProtocol.HTTP,
+    val endpointGroupCreator: () -> Either<Lazy<EndpointGroup>, URI> = {
+        throw IllegalStateException("endpoint/endpoint_group or URI must be set")
+    },
+    val ioThreadsCount: Int? = null,
+    val clientFactoryBuilder: () -> ClientFactoryBuilder = { ClientFactory.builder() },
+    val clientFactoryBuilderCustomizer: ClientFactoryBuilder.() -> ClientFactoryBuilder = { this },
+    val clientOptionsBuilder: () -> ClientOptionsBuilder = { ClientOptions.builder() },
+    val clientOptionsBuilderCustomizer: ClientOptionsBuilder.() -> ClientOptionsBuilder = { this },
+    val rateLimitingDecorator: Function<HttpClient, AutoCloseableHttpClient<*>>? = null
+)
+
+internal abstract class BaseHttpClientBuilderImpl<out HttpClientBuilderT : BaseHttpClientBuilder<HttpClientBuilderT>>(
+    protected val baseBuilderState: BaseHttpClientBuilderState
+) : BaseHttpClientBuilder<HttpClientBuilderT> {
+
+    companion object : Logging
+
+    private lateinit var clientName: String
+
+    internal val lazyClientName: () -> String
+        get() = baseBuilderState.clientNameCreator
+
+    override fun setClientName(clientName: String) = copyOfThisBuilder(
```

Comment:
Motivation here is to allow library clients to reuse some base builders so that they can declare sceleton builder and then specify per client settings. E.g.:

1. lib user could declare base builder with such features as retrying (common retry rule and max retry attempts), requests and connections profiling (common profiler object)
2. then lib user specify different client names (will affect metric names), endpoint groups, timeouts, etc.